go-controller/pkg/ovn/namespace.go

```diff
@@ -244,48 +250,49 @@ func (oc *Controller) updateNamespace(old, newer *kapi.Namespace) {
 	annotation = newer.Annotations[routingExternalGWsAnnotation]
 	oldAnnotation = old.Annotations[routingExternalGWsAnnotation]
 	if annotation != oldAnnotation {
-		var stderr string
-		var err error
-		for podIP, gwToGr := range nsInfo.podExternalRoutes {
-			for gw, gr := range gwToGr {
-				mask := GetIPFullMask(podIP)
-				_, stderr, err = util.RunOVNNbctl("--", "--if-exists", "--policy=src-ip",
-					"lr-route-del", gr, podIP+mask, gw)
+		// if old gw annotation was empty, new one must not be empty, so we should remove any per pod SNAT
+		if oldAnnotation == "" {
+			if config.Gateway.DisableSNATMultipleGWs && (len(nsInfo.routingExternalGWs) != 0 || len(nsInfo.routingExternalPodGWs) != 0) {
+				existingPods, err := oc.watchFactory.GetPods(old.Name)
 				if err != nil {
-					klog.Errorf("Unable to delete src-ip route to GR router, stderr:%q, err:%v", stderr, err)
-				} else {
-					delete(nsInfo.podExternalRoutes, podIP)
+					klog.Errorf("Failed to get all the pods (%v)", err)
 				}
+				for _, pod := range existingPods {
+					logicalPort := podLogicalPortName(pod)
+					portInfo, err := oc.logicalPortCache.get(logicalPort)
+					if err != nil {
+						klog.Warningf("Unable to get port %s in cache for SNAT rule removal", logicalPort)
+					} else {
+						oc.deletePerPodGRSNAT(pod.Spec.NodeName, portInfo.ips)
+					}
+				}
+			}
+		} else {
+			oc.deleteGWRoutesForNamespace(nsInfo)
+		}
+		exGateways, err := parseRoutingExternalGWAnnotation(annotation)
+		if err != nil {
+			klog.Error(err.Error())
+		} else {
+			err = oc.addExternalGWsForNamespace(exGateways, nsInfo, old.Name)
+			if err != nil {
+				klog.Error(err.Error())
 			}
 		}
-		nsInfo.routingExternalGWs = nil
-		if annotation != "" {
-			nsInfo.routingExternalGWs, err = parseRoutingExternalGWAnnotation(annotation)
+		// if new annotation is empty, exgws were removed, may need to add SNAT per pod
+		// check if there are any pod gateways serving this namespace as well
+		if annotation == "" && len(nsInfo.routingExternalPodGWs) == 0 {
```

Comment:
Doesn't this depend on `config.Gateway.DisableSNATMultipleGWs` too? (See above re helper functions...)

Reply:
yes it does! nice catch. will fix it